### ISSUE TITLE
Rework CLI options for KML/COLLADA/glTF export

### ISSUE THREAD

**Issue by clausnagel** [3dcitydb/importer-exporter#150]
The following points have been raised in the discussion of #147:

- [x]  merge `--display-mode` and `--visible-from` into one option
- [x] add options for the elevation settings for KML/COLLADA/glTF exports
- [x] add explicit CLI options for the collada2gltf converter and skip `-O`
was: ~~clarify how to pass parameters to the collada2gltf converter~~

Check out the `export-vis-cli` branch to follow and test the changes.

**Comment by clausnagel** [comment]
@BWibo, I have checked your issue with passing parameters to the collada2gltf converter using @-files. I can confirm that some notations like, for example, `-O '--binary=true','--dracoCompression=true'` work on the CLI but not when using @-files. This example however worked for me after omitting the single quotes `'` in the @-file.

Well, this is rather confusing for users. So, my proposal is to drop the `-O` option and to define specific options like `--gltf-version` and `--gltf-draco-compression`. Opinions?

**Comment by clausnagel** [comment]
Ok, just learned that the parsing of option names _should_ just work fine in picocli as illustrated in the [online documentation](https://picocli.info/#_option_names_as_option_values). But it does not work. Seems to be a bug in picocli. Just checking whether I can work around the bug, so that `-O` can be kept.

**Comment by yaozhihang** [comment]
I like your proposal of dropping the `-O` option and introducing three options `--gltf-version`, `--gltf-binary`, and  `--gltf-draco-compression` instead. This makes the CLI interface clearer and easier to understand.

**Comment by clausnagel** [comment]
I merged the `--visible-from` option into the `--display-mode` option.

I removed the `-O` option and added `--gltf-version`, `--gltf-embed-textures`, `--gltf-binary` and `--gltf-draco-compression` as explicit options to control the behaviour of the collada2gltf converter. Note that the CLI supports abbreviated option names as explained in the [picocli online documentation](https://picocli.info/#_abbreviated_options_and_subcommands). Thus, for example, you don't have to type the full option name `--gltf-draco-compression` but can also use `--gDC`. Much simpler :-) This works for all options of all commands.

I also added options to control the elevation and mainly followed the proposal of @Son-HNguyen in #139 for this.

The `export-vis` is now a monster command (just print out the help). But, well, the feedback shows we obviously need all those options. So, I am thinking to also add the COLLADA rendering options as proposed by @Son-HNguyen in #139 (see `[rendering_options]`). 

The changes are availble in the `export-vis-cli` branch. @yaozhihang, @BWibo, @Son-HNguyen: would be great if you could do some testing.

**Comment by BWibo** [comment]
I just ran the test suite from [here](https://github.com/tum-gis/3dcitydb-importer-exporter-docker/tree/export-vis-cli) after building a fresh image from `export-vis-cli`.
I am seeing some issues regarding export of textured models. For some feature type no textures are exported.

This is what I get for following tests on the railway scene:
* [01, tex missing](https://github.com/tum-gis/3dcitydb-importer-exporter-docker/blob/export-vis-cli/share/config/tests/export-vis/export-vis-gltfOpts.txt)
* [02, tex missing](https://github.com/tum-gis/3dcitydb-importer-exporter-docker/blob/export-vis-cli/share/config/tests/export-vis/export-vis-gltfOpts-02.txt)

![err](https://user-images.githubusercontent.com/7643434/101194372-35a89b00-365e-11eb-90ce-84945df3ac61.png)

* [03, working](https://github.com/tum-gis/3dcitydb-importer-exporter-docker/blob/export-vis-cli/share/config/tests/export-vis/export-vis-gltfOpts-03.txt)

![fine](https://user-images.githubusercontent.com/7643434/101195040-2249ff80-365f-11eb-9a70-2f9305169b8d.png)

It seems it is working when using `--gltf-embed-textures`.

**Comment by clausnagel** [comment]
This also happens when using the GUI. @Son-HNguyen, @yaozhihang, do you have an idea?

**Comment by clausnagel** [comment]
Ok, I could track down the issue to the `KMLExporterManager` class.

In lines 612-616, the COLLADA model is converted to glTF using the collada2gltf converter. This conversion happens _before_ the texture images are written to the file system.

```java
// ----------------- create glTF without embedded textures-----------------
boolean exportGltfV1 = config.getKmlExportConfig().isExportGltfV1();
if (config.getKmlExportConfig().isCreateGltfModel() && !config.getKmlExportConfig().isEmbedTexturesInGltfFiles()) {
    convertColladaToglTF(colladaBundle, buildingDirectory, colladaModelFile, gltfModelFile, exportGltfV1);
}	
```

If the COLLADA model references texture images, then obviously the converter creates 0-byte texture files in the file system.

As next step, the `KMLExporterManager` wants to write the texture images in lines 629-641:

```java
if (colladaBundle.getTexImages() != null) {
   [snipping some lines here...]

    File imageFile = new File(buildingDirectory, imageFilename);
    if (!imageFile.exists()) // avoid overwriting and access conflicts
       ImageIO.write(texImage, imageType, imageFile);					
}
```

The issue is the `if` statement in line 638. Texture images are only written to file if they do not exist already. Now, since the collada2gltf converter created 0-byte texture files in the above step, the images in fact do exist and are not written. Thus, the 0-byte texture files remain in the file system and lead to untextured models when viewing them in Google Earth.

I think a simple fix would be to swap the order of steps: 1) write texture images to the file system, 2) convert the COLLADA model to glTF. I tested this locally and it worked fine. But I don't know whether this leads to side effects.

@yaozhihang and @Son-HNguyen, could you please take over from here?

**Comment by yaozhihang** [comment]
@clausnagel thanks for the investigation. I see no side effects after swapping the two code blocks. While fixing this issue I catched another problem: when the binary-export option is enabled, no .glb files are exported on my Windows 10 OS. Could you @Son-HNguyen @clausnagel also reproduce it?

**Comment by clausnagel** [comment]
No, I cannot confirm this. I am also running on Windows 10, but .glb files are created during export. Do you mind sharing your config files, so that I can test with the exact same settings?

**Comment by clausnagel** [comment]
@BWibo, I fixed the texture export issue (https://github.com/3dcitydb/importer-exporter/commit/96248b9a78d1ac3cc7d53620a64132807ef2eeea) and your three examples now work correctly on my machine.

**Comment by yaozhihang** [comment]
This is my config file (all db connections info have been removed). 

[impexp-config.zip](https://github.com/3dcitydb/importer-exporter/files/5651079/impexp-config.zip)

**Comment by BWibo** [comment]
I can confirm all test are now working using the docker image.

**Comment by yaozhihang** [comment]
The CLI option name and description looks incorrect here
https://github.com/3dcitydb/importer-exporter/blob/export-vis-cli/impexp-client/src/main/java/org/citydb/cli/options/vis/ColladaOption.java#L43

**Comment by clausnagel** [comment]
Hi @yaozhihang, the boolean flag `surfaceNormals` is initialized with true per default (see `defaultValue = "true"`). Only if the option `--no-surface-normals` is specified on the command line, it will be turned to false and no surface normals are generated. So I think option name and description are consistent.

**Comment by yaozhihang** [comment]
Thanks for the explanation. 
I Just have another thought: Could we use  `<rows,columns | sideLength>` instead of `<rows,columns | auto>` for the option `--bbox-tiling`? Then the CLI will also offer the possibility to set tile size for the automatic tiled export.

**Comment by clausnagel** [comment]
Yes, good idea, thanks. This way, you must always provide a value for the side length because the default value (125m) is not applied implicitly anymore. Just wanted to confirm that this is ok. What do you think?

**Comment by yaozhihang** [comment]
I think it is OK. Alternatively, the option could be `<rows,columns | auto,sideLength>`, wenn sideLength is not given, then the default value (125m) will be applied?

**Comment by clausnagel** [comment]
added with 0ad829f3d29039be659f40e133ce9e5232b4c2fb

**Comment by clausnagel** [comment]
> While fixing this issue I catched another problem: when the binary-export option is enabled, no .glb files are exported on my Windows 10 OS. Could you @Son-HNguyen @clausnagel also reproduce it?

Fixed with updating to version 2.1.4 of the collada2gltf converter (89a163d)

**Comment by clausnagel** [comment]
Merged with #151. Closing this issue.
